{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "2_Generating_neighbourhood_and_network_files.ipynb",
      "provenance": [],
      "collapsed_sections": [
        "ed4wCDm2MNy-",
        "geUuPzWHbJ5Q",
        "CJqt287KXDJY",
        "6ofUg8o8blHs"
      ],
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/tqiaowen/LSTrAP-Cloud/blob/master/2_network.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "JGW9c4XmLGL-",
        "colab_type": "text"
      },
      "source": [
        "#Generating Neighbourhood and Network Files\n",
        "This script is based on the original download script found [here](https://github.com/wirriamm/CoNeGC/blob/master/2_Generating_neighbourhood_and_network_files_30Jan.ipynb), obtained on 03 February 2020.\n",
        "\n",
        "### What does this script do?\n",
        "This script will help you perform quality control of the RNA-seq data downloaded by the previous script and generate the files necessary to visualise the co-expression network on this page and also offline on [Cytoscape Desktop](https://cytoscape.org/download.html).\n",
        "\n",
        "### What you will require\n",
        "1. A folder to save the outputs from this script\n",
        "2. Folder created from the previous download script containing the kallisto outputs, download report \n",
        "3. Mercator output (for plants) or file in the mercator output format ([example](https://github.com/tqiaowen/LSTrAP-Cloud/blob/master/examples/mercator_non-plant.txt)) for non-plant organims.<br>NOTE: Gene IDENTIFIERS should be in **lowercase**. File provided should be a **tab-delimited** text file (i.e. '.txt').\n",
        "\n",
        "### Expected outputs\n",
        "1. Gene expression matrix<br>\n",
        "  Contains the tpm values of genes from experiments that passed specified threshold.\n",
        "1. Neighbourhood file<br>\n",
        "   contains information about the neighbourhood of the gene of interest such as geneID, PCC value and descriptions from mercator\n",
        "1. Network file<br>\n",
        "   contains information between gene pairs and their corresponding PCC values\n",
        "1. HTML file<br>\n",
        "   This file can be open standalone in a brower (tested on Chrome) and is the same file used to display the network in this notebook.<br>\n",
        "   **TIP**: Options to download the default network as png and JSON formats are available. The JSON network can be further modified in [Cytoscape Desktop](https://cytoscape.org/download.html).\n",
        "\n",
        "### Bonus tips?\n",
        "1. Location of files saved is in your specified output folder. In addition, we have also printed the location after the files are created. Look out for them after you have ran the cells!\n",
        "2. Want to look at another gene with the same TPM file? **Avoid re-compiling the TPM file** by doing the following steps!<br>\n",
        "  2.1 Edit the description of gene to be analysed in step 2.9<br>\n",
        "  2.2 Re-run cells from 2.9 to 2.14<br>\n",
        "3. If you encoutner any issues, please raise an issue on [Github](https://github.com/tqiaowen/LSTrAP-Cloud) or contact us [here](mailto:qiaowen001@e.ntu.edu.sg)."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "uWINitk0YF5R",
        "colab_type": "code",
        "cellView": "form",
        "colab": {}
      },
      "source": [
        "#@markdown 2.1 Mount Google Drive and import python modules\n",
        "\n",
        "#Mount Google Drive\n",
        "from google.colab import drive\n",
        "drive.mount('/content/gdrive')\n",
        "!rm -rf /content/sample_data\n",
        "\n",
        "#Import modules\n",
        "import os\n",
        "import time\n",
        "from datetime import datetime as dt\n",
        "import scipy.stats as stats\n",
        "import json\n",
        "import math\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "import seaborn as sns\n",
        "import matplotlib.pyplot as plt\n",
        "import IPython\n",
        "from IPython.display import display, HTML\n",
        "from google.colab import widgets #https://colab.research.google.com/notebooks/widgets.ipynb\n",
        "\n",
        "if not os.path.exists(\"/content/MapmanBin_colorsc.txt\"):\n",
        "  !wget https://raw.githubusercontent.com/tqiaowen/LSTrAP-Cloud/master/Dependencies/MapmanBin_colorsc.txt"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "w7I6RbF7YQnf",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "#@title ####2.2 Input form {display-mode: \"form\"}\n",
        "#@markdown Please use absolute paths for the files. Tip: After mounting your Google Drive, right click on the file and copy path.<br>\n",
        "#@markdown <html><div><img src=\"https://github.com/tqiaowen/LSTrAP-Cloud/blob/master/img/copypath.png?raw=true\" alt=\"Copy Path\" width=414 height=250></div></html>\n",
        "\n",
        "#@markdown ---\n",
        "#@markdown Please specify the directory you want your <b>output</b> to be saved.<br> <i>Eg. \"/content/gdrive/My Drive\"</i>\n",
        "outdir = '' #@param {type: 'string'}\n",
        "#@markdown ---\n",
        "\n",
        "#@markdown Please specify the directory containing the kallisto files you want to analyse.<br> <i>Eg. \"/content/gdrive/My Drive/Nta/2020-01-31_07/\"</i>\n",
        "indir = '' #@param {type: 'string'}\n",
        "#@markdown ---\n",
        "\n",
        "#@markdown Abbreviation to be used for species.<br> <i>Eg. For _Nicotiana tabacum_, it can be \"Nta\"</i>\n",
        "abv = '' #@param {type: 'string'}\n",
        "#@markdown ---\n",
        "\n",
        "#@markdown Location of <a href=\"https://www.plabipd.de/portal/mercator4\" target=\"_blank\">mercator</a> output containing corresponding MapMan bins for genes in CDS.<br> <i>Eg. \"/content/gdrive/My Drive/Nta/Mercator_Nta.txt\"</i><br>\n",
        "#@markdown If you are using a non-plant organism, please provide a file based on the mercator format available <a href=\"https://github.com/tqiaowen/LSTrAP-Cloud/blob/master/examples/mercator_non-plant.txt\" target=\"_blank\">here</a>. Note: Gene IDENTIFIER should be in **lowercase**.\n",
        "merloc = '' #@param {type: 'string'}\n",
        "#@markdown ---\n",
        "\n",
        "#@markdown Location of Download report.<br> <i>Eg. \"/content/gdrive/My Drive/Nta/2020-01-31_07/Download_report_Nta_2020-01-31.txt\"</i>\n",
        "reploc = '' #@param {type: 'string'}\n",
        "#@markdown ---\n",
        "\n",
        "# #@markdown Description of gene to be analysed (must be the same as the one found in the CDS file)<br><i>Eg. \"lcl|NW_015787229.1_cds_XP_016497259.1_1\"</i>\n",
        "# goi = '' #@param {type: 'string'}\n",
        "# #@markdown ---\n",
        "\n",
        "#Initialise variables\n",
        "working_dir_path = outdir + \"/\"\n",
        "working_dir_path_ter = outdir.replace(\" \", \"\\ \")\n",
        "execution_dir_path = indir + \"/\"\n",
        "execution_dir_path_ter = indir.replace(\" \", \"\\ \")\n",
        "os.chdir(execution_dir_path)\n",
        "goi = goi.lower()\n",
        "\n",
        "tpm_table_file_name = \"compiled_tpm_table_\" + abv + \".txt\""
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "YSF6Cvf7WQxy",
        "colab_type": "text"
      },
      "source": [
        "### Quality control"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "jbtEQ78b4HWq",
        "colab_type": "code",
        "cellView": "form",
        "colab": {}
      },
      "source": [
        "#@markdown 2.3 Preparing data for visualisation\n",
        "\n",
        "RunIDs = []\n",
        "pseudoaligned = []\n",
        "processed = []\n",
        "p_pseudoaligned = []\n",
        "log10processed = []\n",
        "p_genes_mapped = []\n",
        "\n",
        "with open(reploc, \"r\") as report:\n",
        "  report.readline()\n",
        "  for line in report:\n",
        "    if \"#\" not in line and line.split(\"\\t\")[2] == \"Downloaded\": #only successfully downloaded files will be considered\n",
        "      line_con = line.strip().split(\"\\t\")\n",
        "      if int(line_con[6]) != 0:\n",
        "        log10processed.append(np.log10(int(line_con[6])))\n",
        "      else:\n",
        "        log10processed.append(0)\n",
        "      RunIDs.append(line_con[0])\n",
        "      processed.append(int(line_con[6]))\n",
        "      pseudoaligned.append(int(line_con[7]))\n",
        "      p_pseudoaligned.append(float(line_con[8]))\n",
        "      p_genes_mapped.append(float(line_con[9].split(\"%\")[0]))\n",
        "    \n",
        "qc_df = pd.DataFrame({'RunID': RunIDs, 'Reads Pseudoaligned':pseudoaligned, 'Reads Processed':processed, '% pseudoaligned': p_pseudoaligned, 'log10(Processed Reads)': log10processed, '% genes mapped': p_genes_mapped})\n",
        "\n",
        "tot_genes = 0\n",
        "with open(execution_dir_path+RunIDs[0]+\"/abundance.tsv\",\"r\") as samp:\n",
        "  samp_con = [x.strip() for x in samp.readlines()]\n",
        "  tot_genes = len(samp_con) -1"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "DhKrqzaQMh6j",
        "colab_type": "text"
      },
      "source": [
        "####QC Inspection and Visualisation"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "L9u8jI6TD4g5",
        "colab_type": "code",
        "cellView": "form",
        "colab": {}
      },
      "source": [
        "#@markdown 2.4 Display Quality Control Table\n",
        "from IPython.display import Javascript\n",
        "display(Javascript('''google.colab.output.setIframeHeight(0, true, {maxHeight: 500})'''))\n",
        "display(HTML(\"Total number of genes in organism: \" + str(tot_genes)))\n",
        "\n",
        "tab = widgets.TabBar([\"Sorted by reads processed\", \"Sorted by % pseudoaligned\", \"Sorted by % genes_mapped\"], location = \"top\")\n",
        "columns = [\"Reads Processed\", \"% pseudoaligned\", \"% genes mapped\"]\n",
        "for sort_column in columns:\n",
        "  new_qc_df = qc_df.sort_values(by=[sort_column], inplace=False, ascending=False)\n",
        "  with tab.output_to(columns.index(sort_column), select=False):\n",
        "    pd.set_option('display.max_rows', 200)\n",
        "    display(HTML(new_qc_df.to_html()))"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "m7L9ArUPSuUs",
        "colab_type": "code",
        "cellView": "form",
        "colab": {}
      },
      "source": [
        "#@markdown 2.5 Scatter plot of log10( number of processed reads ) against % of targets pseudoaligned in each experiment\n",
        "sns.set(style=\"darkgrid\")\n",
        "sns.set(font_scale=1.4)\n",
        "qc_dfa = pd.DataFrame({'% pseudoaligned': p_pseudoaligned, 'log10(Processed Reads)': log10processed})\n",
        "g = sns.jointplot('% pseudoaligned', 'log10(Processed Reads)', data = qc_dfa, kind=\"scatter\",\n",
        "                  xlim=(0, 100), ylim=(0, 10), color=\"m\", height=10, marginal_kws=dict(bins=20, rug=True))\n",
        "#plt.subplots_adjust(top=0.9)\n",
        "#g.fig.suptitle(\"Log10( number of processed reads ) vs % of targets pseudoaligned\")"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "UZ6A3lIefQo6",
        "colab_type": "code",
        "cellView": "form",
        "colab": {}
      },
      "source": [
        "#@markdown 2.6 Scatter plot of % of genes mapped against % of targets pseudoaligned in each experiment\n",
        "\n",
        "sns.set(style=\"darkgrid\")\n",
        "sns.set(font_scale=1.4)\n",
        "qc_dfb = pd.DataFrame({'% pseudoaligned': p_pseudoaligned, '% genes mapped': p_genes_mapped})\n",
        "g = sns.jointplot('% pseudoaligned', '% genes mapped', data = qc_dfb, kind=\"scatter\",\n",
        "                  xlim=(0, 100), ylim=(0, 100), color=\"m\", height=10, marginal_kws=dict(bins=40, rug=True))"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "-nH8Ch0GWoNT",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "#@title ####2.7 Determine Quality Control Cutoff {display-mode: \"form\", run: \"auto\"}\n",
        "#@markdown Please assess the quality of the downloaded files using the tables and graphs above. By adjusting the sliders below, only RunIDs (experiments) with values above the threshold will be selected for further analysis.\n",
        "\n",
        "p_pseudoaligned_threshold = 40  #@param {type: \"slider\", min: 0, max: 100}\n",
        "\n",
        "#@markdown Enter the log<sub>10</sub> of reads processed instead of absolute number.\n",
        "reads_pseudoalinged_threshold = 6 #@param {type: \"slider\", min: 0, max: 10, step: 0.1}\n",
        "\n",
        "p_genes_mapped_threshold = 50 #@param {type: \"slider\", min: 0, max: 100}\n",
        "\n",
        "#title Collect RunIDs above threshold\n",
        "accepted_RunIDs = []\n",
        "ra = qc_df.loc[:,\"% pseudoaligned\"]\n",
        "rp = qc_df.loc[:,\"Reads Pseudoaligned\"]\n",
        "gm = qc_df.loc[:,\"% genes mapped\"]\n",
        "\n",
        "for i in range(len(ra)):\n",
        "  if ra[i] >= p_pseudoaligned_threshold and rp[i] >= reads_pseudoalinged_threshold and gm[i] >= p_genes_mapped_threshold:\n",
        "    accepted_RunIDs.append(qc_df.loc[i,\"RunID\"])\n",
        "\n",
        "print(\"Number of accepted Run IDs:\"+ str(len(accepted_RunIDs)) + \"/\" + str(len(ra)))\n",
        "\n",
        "#to_recompile_tpm_values = False #@param {type:\"boolean\"}"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "tMQotbx0WyKo",
        "colab_type": "text"
      },
      "source": [
        "###Data Extraction"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "gG2-d40Hzfgr",
        "colab_type": "code",
        "cellView": "form",
        "colab": {}
      },
      "source": [
        "#@markdown 2.8 Compile TPM values of selected RunIDs\n",
        "#@markdown <html>&nbsp&nbsp&nbsp&nbsp2.8.1 Functions to compile TPM values</html>\n",
        "#Functions#############################################################\n",
        "def compile_tpm(accepted_RunIDs, tpm_table_file_name):\n",
        "  #Iterate through the list of selected RunIDs\n",
        "  for RunID in accepted_RunIDs:\n",
        "    RunID_file_path = execution_dir_path + RunID + \"/\"\n",
        "    if os.path.exists(RunID_file_path) == False:\n",
        "      print(RunID + \": RunID directory not found.\")\n",
        "    else:\n",
        "      try:\n",
        "        abun = pd.read_csv(RunID_file_path + \"abundance.tsv\", sep='\\t', header=0, engine=\"python\") #opens abundance.tsv as df\n",
        "        if RunID == accepted_RunIDs[0]:\n",
        "          master_tpm = abun[[\"target_id\"]].copy()\n",
        "        if abun.isnull().values.any() != True:\n",
        "          master_tpm.insert(1, RunID, abun[\"tpm\"]) #adds tpm column to collection of tpm values\n",
        "      except:\n",
        "        print(\"Unable to compile \" + RunID)\n",
        "  master_tpm[\"target_id\"] = master_tpm[\"target_id\"].apply(lambda x: x.lower()) #convert gene IDs to lowercase to match mercator results\n",
        "  master_tpm.to_csv(working_dir_path + tpm_table_file_name, index=False, sep=\"\\t\")\n",
        "  print(\"TPM table created at:\" + working_dir_path + tpm_table_file_name + \".\\n\")\n",
        "\n",
        "######################################################################\n",
        "#@markdown <html>&nbsp&nbsp&nbsp&nbsp2.8.2 Compiling TPM values <small><strong>(This step may take a while to run)</small></strong></html>\n",
        "#@markdown \n",
        "tpm_table_file_name = \"compiled_tpm_table_\" + abv + \".txt\"\n",
        "# dict_of_genes = compile_tpm(accepted_RunIDs, tpm_table_file_name)\n",
        "compile_tpm(accepted_RunIDs, tpm_table_file_name)"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "fOrezZOEXGqY",
        "colab_type": "code",
        "cellView": "form",
        "colab": {}
      },
      "source": [
        "#@markdown 2.9 Network options {display-mode: \"form\", run: \"auto\"}\n",
        "#@markdown <br><br>Description of gene to be analysed (must be the same as the one found in the CDS file)<br><i>Eg. \"lcl|NW_015787229.1_cds_XP_016497259.1_1\"</i>\n",
        "goi = '' #@param {type: 'string'}\n",
        "cutoff = 0.70  #@param {type:\"slider\", min:0, max:1, step:0.01}\n",
        "neighbourhood_size = 51 #@param {type: \"slider\", min:1, max:201, step:1}\n",
        "#@markdown Please uncheck if your organism is not a plant.</strong>\n",
        "is_a_plant = True #@param {type: \"boolean\"}\n",
        "\n",
        "#to_recompile_files = False #@param {type:\"boolean\"}"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "fAQs0S6nXNu7",
        "colab_type": "code",
        "cellView": "form",
        "colab": {}
      },
      "source": [
        "#@markdown 2.10 Neighbourhood and network files\n",
        "#@markdown <html>&nbsp&nbsp&nbsp&nbsp2.10.1 Functions to generate neighbourhood and network files</html>\n",
        "#Functions#############################################################\n",
        "\n",
        "def extract_mercator(mercator_file_path):\n",
        "  #Stores gene description\n",
        "  with open(mercator_file_path, 'r') as mercator_file:\n",
        "    mercator_file.readline() #header\n",
        "    mercator_lines = mercator_file.readlines()\n",
        "\n",
        "  description_dict = {}\n",
        "\n",
        "  for line in mercator_lines:\n",
        "    line_list = line.strip().split('\\t')\n",
        "    geneID = line_list[2].strip('\\'')\n",
        "    bin_code = line_list[0].strip('\\'')\n",
        "    bin_name = line_list[1].strip('\\'')\n",
        "    gene_description = line_list[3].strip('\\'')\n",
        "    if geneID != '':\n",
        "      description_dict[geneID] = [bin_code, bin_name, gene_description]\n",
        "\n",
        "  return description_dict\n",
        "\n",
        "######################################################################\n",
        "\n",
        "def read_tpm_table(tpm_table_file_path):\n",
        "  #Read TPM Table\n",
        "  compiled_tpm_table = open(tpm_table_file_path, \"r\")\n",
        "  #Read header and store RunIDs, use index of RunID to identify. \n",
        "  RunID_list = compiled_tpm_table.readline().strip().split()[1:]\n",
        "\n",
        "  #Read each row for each gene\n",
        "  #[genes, ...]\n",
        "  #[[tpm of gene 1], [tpm of gene 2], ...]\n",
        "  genes_tpm_rows = compiled_tpm_table.readlines() #list of rows\n",
        "  genes_list = [line.strip().split('\\t')[0] for line in genes_tpm_rows]\n",
        "  tpm_list_of_lists = [line.strip().split('\\t')[1:] for line in genes_tpm_rows]\n",
        "  compiled_tpm_table.close()\n",
        "\n",
        "  return RunID_list, genes_list, tpm_list_of_lists\n",
        "\n",
        "# ######################################################################\n",
        "\n",
        "# def verify_guide_genes(list_of_guide_genes, genes_list):\n",
        "#   #Check if guide genes inserted in input section is a GeneID present in genes_list\n",
        "#   for guide_gene in list_of_guide_genes:\n",
        "#     if guide_gene not in genes_list:\n",
        "#       print(\"The gene \" + guide_gene + \" is invalid.\")\n",
        "#       list_of_guide_genes.remove(guide_gene)\n",
        "  \n",
        "#   return list_of_guide_genes\n",
        "\n",
        "######################################################################\n",
        "\n",
        "def make_neighbourhood_file(GoI, neighbourhood_file_path, genes_list, tpm_list_of_list, description_dict):\n",
        "  #1 Create neightbourhood file output for this guide gene\n",
        "  neighbourhood_file = open(neighbourhood_file_path, 'w+')\n",
        "  neighbourhood_file.write('Neighbourhood file for ' + GoI + '\\n')\n",
        "  neighbourhood_file.write('Gene ID\\tPCC Value\\tp-value\\tMapman Bin Code\\tMapman Bin Name\\tDescription\\n')\n",
        "\n",
        "  GoI_index = genes_list.index(GoI)\n",
        "  tpm_GoI = [float(tpm) for tpm in tpm_list_of_lists[GoI_index]]\n",
        "\n",
        "  #Generate a list of list [[gene ID, PCC, p-value], [], ...]\n",
        "  PCC_output = []\n",
        "\n",
        "  #Pair GoI with every other gene in genes_list\n",
        "  for i in range(len(genes_list)):\n",
        "    tpm_gene2 = [float(tpm) for tpm in tpm_list_of_lists[i]]\n",
        "    PCC_value, p_value = stats.pearsonr(tpm_GoI, tpm_gene2)\n",
        "    if PCC_value > cutoff: #Only append pairs with PCC > cutoff\n",
        "      PCC_output.append([genes_list[i], PCC_value, p_value])\n",
        "\n",
        "  #Sort by PCC\n",
        "  PCC_sorted = sorted(PCC_output, key = lambda x: x[1], reverse = True)\n",
        "  PCC_top = PCC_sorted[:neighbourhood_size] #Write only the top PCC by neighbourhood size\n",
        "\n",
        "  #Add gene description\n",
        "  #description_dict = {geneID: [bin_code, gene_name, gene_description]}\n",
        "  PCC_top_annotated = []\n",
        "  for result in PCC_top:\n",
        "    geneID = result[0]\n",
        "    result.extend(description_dict[geneID]) #extend with [bin_code, bin name, genedesc]\n",
        "    PCC_top_annotated.append(result) #[gene ID, PCC, p-value, bin_code, bin name, genedesc]\n",
        "\n",
        "  #write to neighbourhood file\n",
        "  for result in PCC_top_annotated: \n",
        "    result = [str(e) for e in result]\n",
        "    to_write = '\\t'.join(result) + '\\n' \n",
        "    neighbourhood_file.write(to_write)\n",
        "\n",
        "  neighbourhood_file.close()\n",
        "\n",
        "  print(\"Neighbourhood file created at: \" + neighbourhood_file_path)\n",
        "\n",
        "######################################################################\n",
        "\n",
        "def create_network(GoI, neighbourhood_file_path, network_file_path, genes_list, tpm_list_of_lists):\n",
        "  #Read neightbourhood file output\n",
        "  with open(neighbourhood_file_path, 'r') as neighbourhood_file:\n",
        "    neighbourhood_file.readline() #Title\n",
        "    neighbourhood_file.readline() #Headers\n",
        "    neighbourhood_lines = neighbourhood_file.readlines()\n",
        "\n",
        "  #Collect top genes including GoI itself\n",
        "  network_genes = [line.strip().split('\\t')[0] for line in neighbourhood_lines]\n",
        "\n",
        "  #Create network file output\n",
        "  network_file = open(network_file_path, 'w+')\n",
        "  network_file.write('Network file for ' + GoI + '\\n')\n",
        "  network_file.write('Gene ID 1\\tGene ID 2\\tPCC Value\\tp-value\\n')\n",
        "\n",
        "  #Iterate every gene\n",
        "  for gene1 in network_genes:\n",
        "    network_genes_index = network_genes.index(gene1) #for iterating through only genes in neighbourhood file\n",
        "    genes_list_index1 = genes_list.index(gene1) #find index of gene in genes_list from container extracted by read_tpm_table function\n",
        "    #Pair gene with every other gene below it in the list\n",
        "    for i in range((network_genes_index + 1), len(network_genes)): #gene2 is iterated through all the genes below gene1 \n",
        "      gene2 = network_genes[i]\n",
        "      tpm_gene1 = [float(Exp) for Exp in tpm_list_of_lists[genes_list_index1]]\n",
        "      genes_list_index2 = genes_list.index(gene2)\n",
        "      tpm_gene2 = [float(Exp) for Exp in tpm_list_of_lists[genes_list_index2]]\n",
        "      PCC_value, p_value = stats.pearsonr(tpm_gene1, tpm_gene2)\n",
        "\n",
        "      if math.isnan(PCC_value):\n",
        "        PCC_value, p_value = 0, 0\n",
        "\n",
        "      if cutoff <= PCC_value <= 1.0:\n",
        "        to_write = gene1 + '\\t' + gene2 + '\\t' + str(PCC_value) + '\\t' + str(p_value) + '\\n' \n",
        "        network_file.write(to_write)\n",
        "\n",
        "  network_file.close()\n",
        "  print(\"Network file created at: \" + network_file_path)\n",
        "\n",
        "######################################################################\n",
        "\n",
        "#@markdown <html>&nbsp&nbsp&nbsp&nbsp2.10.2 Generate neighbourhood and network files</html>\n",
        "# Extract Mercator Description\n",
        "description_dict = extract_mercator(merloc)\n",
        "\n",
        "# Read TPM table \n",
        "tpm_table_file_path = working_dir_path + tpm_table_file_name \n",
        "RunID_list, genes_list, tpm_list_of_lists = read_tpm_table(tpm_table_file_path) \n",
        "\n",
        "if goi in genes_list:\n",
        "#for guide_gene in list_of_guide_genes:\n",
        "  #1 Create neighbourhood file\n",
        "  neighbourhood_file_path = working_dir_path + 'Neighbourhood_' + goi.replace(\"|\",\"_\") + '.txt'\n",
        " # if (not os.path.exists(neighbourhood_file_path)) or to_recompile_files == True:\n",
        "  make_neighbourhood_file(goi, neighbourhood_file_path, genes_list, tpm_list_of_lists, description_dict)\n",
        " # else:\n",
        " #   print(goi + \": neighbourhood file already present.\")\n",
        "  \n",
        "  #2 Create network file\n",
        "  network_file_path = working_dir_path + 'Network_' + goi.replace(\"|\",\"_\") + '.txt'\n",
        "  # if (not os.path.exists(network_file_path)) or to_recompile_files == True:\n",
        "  create_network(goi, neighbourhood_file_path, network_file_path, genes_list, tpm_list_of_lists) #Requires neighbourhood file to be created first\n",
        "  # else:\n",
        "  #   print(goi + \": network file already present.\")"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "u1XX0GDDa45S",
        "colab_type": "text"
      },
      "source": [
        "##Visualisation of Co-expression Networks"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "XGlsufbIa8O8",
        "colab_type": "code",
        "cellView": "form",
        "colab": {}
      },
      "source": [
        "#@markdown 2.11 Generating HTML file for network visualisation\n",
        "#@markdown <html>&nbsp&nbsp&nbsp&nbsp2.11.1 Functions to generate html file for network visualisation</html>\n",
        "#Functions#############################################################\n",
        "\n",
        "def extract_nodes_and_edges(network_file_path):\n",
        "  '''Opens a network file and return a list of all the edges and nodes within the network'''\n",
        "  with open(network_file_path, 'r') as network_file:\n",
        "    network_file.readline() #title\n",
        "    network_file.readline() #header\n",
        "    entries = [lines.strip().split('\\t') for lines in network_file.readlines()]\n",
        "\n",
        "  list_of_edges = []\n",
        "  for i in range(len(entries)):\n",
        "    if float(entries[i][2]) >= 0.7:\n",
        "      edge = (entries[i][0], entries[i][1])\n",
        "      list_of_edges.append(edge)\n",
        "\n",
        "  list_of_nodes = []\n",
        "  for j in list_of_edges:\n",
        "    if j[0] not in list_of_nodes:\n",
        "      list_of_nodes.append(j[0])\n",
        "    if j[1] not in list_of_nodes:\n",
        "      list_of_nodes.append(j[1])\n",
        "  \n",
        "  return list_of_nodes, list_of_edges\n",
        "\n",
        "######################################################################\n",
        "\n",
        "def style_one_node(GeneID, shape, color):\n",
        "  '''Returns the string of html css style sheet to be added to the index html file'''\n",
        "  node_styling='''\\\n",
        "              {\n",
        "              selector: 'node[id=\"GENEID\"]',\n",
        "              style: {\n",
        "                  shape: 'ENTER_SHAPE',\n",
        "                  'background-color': 'ENTER_COLOR',\n",
        "                  }\n",
        "              },\n",
        "  '''\n",
        "  node_styling = node_styling.replace('GENEID', GeneID)\n",
        "  node_styling = node_styling.replace('ENTER_SHAPE', shape)\n",
        "  node_styling = node_styling.replace('ENTER_COLOR', color)\n",
        "\n",
        "  return node_styling\n",
        "\n",
        "######################################################################\n",
        "\n",
        "def read_neighbourhood_file(neighbourhood_file_path):\n",
        "  with open(neighbourhood_file_path, \"r\") as neighbourhood_file:\n",
        "    neighbourhood_file.readline() #title\n",
        "    neighbourhood_file.readline() #header\n",
        "    lines_list = [line.strip().split(\"\\t\") for line in neighbourhood_file.readlines()]\n",
        "\n",
        "  # gene_info_dict = {GeneId : [bin_code, bin_fullname, description, ##arabiposis ortholog?##]}\n",
        "  gene_info_dict = {}\n",
        "  for line in lines_list:\n",
        "    gene_info_dict[line[0]] = [line[3], line[4], line[5]]\n",
        "  \n",
        "  return gene_info_dict\n",
        "\n",
        "######################################################################\n",
        "\n",
        "def read_color_table(MapmanBin_color_table_path):\n",
        "  '''Returns color_dict = {Bin number: [Bin name, shape, color]}'''\n",
        "  with open(MapmanBin_color_table_path, \"r\") as color_table:\n",
        "    color_table.readline() #header\n",
        "    lines = [line.strip().split(\"\\t\") for line in color_table.readlines()]\n",
        "  \n",
        "  color_dict = {}\n",
        "  for line in lines:\n",
        "    color_dict[int(line[0])] = [line[1], line[2], line[3]]\n",
        "  \n",
        "  return color_dict\n",
        "\n",
        "######################################################################\n",
        "\n",
        "def generate_html_index(list_of_nodes, list_of_edges):\n",
        "  '''Returns a string of a html with javascript code to render the network from list_of_nodes and list_of_edges'''\n",
        "  front_html = '''\\\n",
        "  <!doctype html>\n",
        "  <html>\n",
        "    <head>\n",
        "        <title>Network on Cytoscape</title>\n",
        "        <script src='https://cdnjs.cloudflare.com/ajax/libs/cytoscape/3.9.4/cytoscape.min.js'></script>\n",
        "    </head>\n",
        "\n",
        "    <style>\n",
        "        #cy {\n",
        "            width: 1000px;\n",
        "            height: 950px;\n",
        "            }\n",
        "    </style>\n",
        "\n",
        "    <body style=\"background-color:rgb(194, 229, 243);\">\n",
        "      <div><p>Click to download as: <br><a id=\"pngimage\" href=\"#\" download=\"NetworkImage.png\">Image</a> | <a id=\"cyjson\" href=\"#\" download=\"NetworkImage.json\">JSON</a></p></div>\n",
        "      <div id=\"cy\"></div>\n",
        "      <script>\n",
        "        var cy = cytoscape({\n",
        "          container: document.getElementById('cy'),\n",
        "  '''\n",
        "\n",
        "\n",
        "  mid_html = '''\\\n",
        "          elements: {\n",
        "              nodes: [\\n\n",
        "  '''\n",
        "\n",
        "  #Read neighbourhood files\n",
        "  neighbourhood_file_path = working_dir_path + 'Neighbourhood_' + goi.replace(\"|\",\"_\") + '.txt'\n",
        "  gene_info_dict = read_neighbourhood_file(neighbourhood_file_path) #Need to include arabidopsis orthologs too! (run diamond?)\n",
        "\n",
        "  #Read Mapman Bin to Color table\n",
        "  MapmanBin_color_table_path = \"/content/MapmanBin_colorsc.txt\"\n",
        "  color_dict = read_color_table(MapmanBin_color_table_path)\n",
        "\n",
        "  #get nodes\n",
        "  # { data: { id: 'a', node_type: 'q/t', color: 'x', shape: 'z' } },\n",
        "  if is_a_plant == True:\n",
        "    for j in list_of_nodes:\n",
        "      bin_code = int(gene_info_dict[j][0].split(\".\")[0])\n",
        "      shape, color = color_dict[bin_code][1], color_dict[bin_code][2]\n",
        "      ntype = \"linked\"\n",
        "      if j == list_of_nodes[0]:\n",
        "        ntype = \"query\"\n",
        "        color = \"gold\"\n",
        "        shape = \"vee\"\n",
        "      new_node = \"\\t\"*5 + \"{\\n\" + \"\\t\"*6 + \"data: {\\n\" +\\\n",
        "        \"\\t\"*6 + \"id: \\'\" + j + \"\\',\\n\" +\\\n",
        "        \"\\t\"*6 + \"node_type: \\'\" + ntype + \"\\',\\n\" +\\\n",
        "        \"\\t\"*6 + \"color: \\'\" + color + \"\\',\\n\" +\\\n",
        "        \"\\t\"*6 + \"shape: \\'\" + shape + \"\\'\\n\" + \"\\t\"*6 + \"}\\n\" + \"\\t\"*5 + \"}\"\n",
        "      if list_of_nodes.index(j) != len(list_of_nodes)-1:\n",
        "        new_node += \",\\n\"\n",
        "      else:\n",
        "        new_node += \"\\n\\t\\t\\t\\t],\\n\"\n",
        "      mid_html += new_node\n",
        "  else:\n",
        "    for j in list_of_nodes:\n",
        "      shape, color = \"triangle\", \"Silver\"\n",
        "      ntype = \"linked\"\n",
        "      if j == list_of_nodes[0]:\n",
        "        ntype = \"query\"\n",
        "        color = \"Gold\"\n",
        "        shape = \"vee\"\n",
        "      new_node = \"\\t\"*5 + \"{\\n\" + \"\\t\"*6 + \"data: {\\n\" +\\\n",
        "        \"\\t\"*6 + \"id: \\'\" + j + \"\\',\\n\" +\\\n",
        "        \"\\t\"*6 + \"node_type: \\'\" + ntype + \"\\',\\n\" +\\\n",
        "        \"\\t\"*6 + \"color: \\'\" + color + \"\\',\\n\" +\\\n",
        "        \"\\t\"*6 + \"shape: \\'\" + shape + \"\\'\\n\" + \"\\t\"*6 + \"}\\n\" + \"\\t\"*5 + \"}\"\n",
        "      if list_of_nodes.index(j) != len(list_of_nodes)-1:\n",
        "        new_node += \",\\n\"\n",
        "      else:\n",
        "        new_node += \"\\n\\t\\t\\t\\t],\\n\"\n",
        "      mid_html += new_node\n",
        "\n",
        "  mid2_html = '''\\\n",
        "              edges: [\\n\n",
        "  '''\n",
        "  mid_html += mid2_html\n",
        "  #get edges\n",
        "  for i in range(len(list_of_edges)):\n",
        "    ID = list_of_edges[i][0] + \" | \" + list_of_edges[i][1]\n",
        "    source = list_of_edges[i][0]\n",
        "    target = list_of_edges[i][1]\n",
        "    new_edge = \"\\t\"*5 + \"{\\n\" + \"\\t\"*6 + \"data: {\\n\" +\\\n",
        "      \"\\t\"*7 + \"id: \\'\" + ID + \"\\',\\n\" +\\\n",
        "      \"\\t\"*7 + \"source: \\'\" + source + \"\\',\\n\" +\\\n",
        "      \"\\t\"*7 + \"target: \\'\" + target + \"\\'\\n\" + \"\\t\"*6 + \"}\\n\" + \"\\t\"*5 + \"}\"\n",
        "    if i != (len(list_of_edges)-1):\n",
        "      new_edge += \",\\n\"\n",
        "    else:\n",
        "      new_edge = new_edge + \"\\n\\t\\t\\t\\t]\\n\"\n",
        "    mid_html += new_edge\n",
        "\n",
        "  back_html = '''\\\n",
        "            },\n",
        "            style: [\n",
        "              { \n",
        "              selector: \"node\",\n",
        "              style: {\n",
        "                  shape: 'data(shape)',\n",
        "                  'height': '3px',\n",
        "                  'width': '3px',\n",
        "                  'background-color': 'data(color)',\n",
        "                  label: 'data(id)',\n",
        "                  'border-style':'solid',\n",
        "                  'border-color':'black',\n",
        "                  'border-width':'0.1px',\n",
        "                  'font-size': 1.2,\n",
        "                  'text-margin-y':'-0.4px',\n",
        "                  'text-outline-color':'white',\n",
        "                  'text-outline-width':'0.15px',\n",
        "                  'text-outline-opacity':0.8\n",
        "                  }\n",
        "              },\n",
        "              {\n",
        "              selector: 'edge',\n",
        "              style: {\n",
        "                  'width': '0.1px',\n",
        "                  'line-style':'solid',\n",
        "                  'line-fill':'solid',\n",
        "                  'background-color': '#F5F5F5',\n",
        "                  'opacity':0.4\n",
        "                  }\n",
        "              },\n",
        "              {\n",
        "              selector:'core',\n",
        "              style: {\n",
        "                  'outside-texture-bg-color':'white',\n",
        "                  'outside-texture-bg-opacity':1\n",
        "                  }\n",
        "              },\n",
        "            ],\n",
        "            layout: {\n",
        "              name: 'cose',\n",
        "              animateThreshold: 1,\n",
        "              nodeOverlap: 100,\n",
        "              gravity: 400,\n",
        "              componentSpacing: 100,\n",
        "              edgeElasticity: 100,\n",
        "              minTemp: 1,\n",
        "              initialTemp: 1000,\n",
        "              stop: function() {\n",
        "                  //generate PNG image to display from \"id = pngimage\"\n",
        "                  var png64 = cy.png({bg:'white'});\n",
        "                  document.getElementById('pngimage').setAttribute('href', png64);\n",
        "                  //generate JSON to download for id = cyjson\n",
        "                  var jsonc = JSON.stringify(cy.json(), null, \"\\\\t\");\n",
        "                  document.getElementById('cyjson').setAttribute('href', 'data:text/plain;charset=utf-8,' + encodeURIComponent(jsonc));\n",
        "              }\n",
        "            }\n",
        "        });\n",
        "\n",
        "      </script>\n",
        "    </body>\n",
        "  </html>\n",
        "  '''\n",
        "\n",
        "  index_html = front_html + mid_html + back_html\n",
        "  return index_html\n",
        "\n",
        "######################################################################\n",
        "\n",
        "def save_index_file(index_html, guide_gene):\n",
        "  '''Writes the html/js string into a html file and return the path to the new html file'''\n",
        "  network_html_path = working_dir_path + \"Network_index_\" + guide_gene + '.html'\n",
        "  with open(network_html_path, 'w+') as index_file:\n",
        "    index_file.write(index_html)\n",
        "  print(\"HTML file saved in: \" + network_html_path+ \"\\n\")\n",
        "  return network_html_path\n",
        "\n",
        "######################################################################\n",
        "\n",
        "#@markdown <html>&nbsp&nbsp&nbsp&nbsp2.11.2 Create html files for network visualisation. \n",
        "#<strong>Please uncheck if your organism is not a plant.</strong><br></html>\n",
        "#is_a_plant = True #@param {type: \"boolean\"}\n",
        "\n",
        "#Read network file for nodes and edges\n",
        "list_of_nodes, list_of_edges = extract_nodes_and_edges(working_dir_path + \"Network_\" + goi.replace(\"|\",\"_\") + \".txt\")\n",
        "index_html = generate_html_index(list_of_nodes, list_of_edges) #string\n",
        "index_html = index_html.replace('NetworkImage', goi.replace(\"|\",\"_\")) #Specify Name of png file to be downloaded based on guide gene\n",
        "\n",
        "# Save string into final html file\n",
        "network_html_path = save_index_file(index_html, goi.replace(\"|\",\"_\"))\n",
        "#list_of_network_html_path.append(network_html_path)"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Ry3l-h0Ea76K",
        "colab_type": "code",
        "cellView": "form",
        "colab": {}
      },
      "source": [
        "#@markdown 2.12 Display Network\n",
        "#network_html_path = list_of_network_html_path[0]\n",
        "IPython.display.HTML(filename = network_html_path)"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "_I4GLTWlcEqz",
        "colab_type": "code",
        "cellView": "form",
        "colab": {}
      },
      "source": [
        "#@markdown 2.13 Legend for nodes based on Mapman Bins:\n",
        "\n",
        "nodes_elements = \"\"\n",
        "\n",
        "if is_a_plant == True:\n",
        "  #color_dict = {Bin number: [Bin name, shape, color]}\n",
        "  MapmanBin_color_table_path = \"/content/MapmanBin_colorsc.txt\"\n",
        "  color_dict = read_color_table(MapmanBin_color_table_path)\n",
        "\n",
        "  for key in color_dict:\n",
        "    bin_name, shape, color = color_dict[key][0], color_dict[key][1], color_dict[key][2]\n",
        "    #Add node to elements\n",
        "    # { data: { id: 'a', color: 'x', shape: 'z' } },\n",
        "    new_node = \"\\t\"*5 + \"{\\n\" + \"\\t\"*6 + \"data: {\\n\" +\\\n",
        "        \"\\t\"*6 + \"id: \\'\" + bin_name + \"\\',\\n\" +\\\n",
        "        \"\\t\"*6 + \"color: \\'\" + color + \"\\',\\n\" +\\\n",
        "        \"\\t\"*6 + \"shape: \\'\" + shape + \"\\'\\n\" + \"\\t\"*6 + \"}\\n\" + \"\\t\"*5 + \"},\\n\"\n",
        "    nodes_elements += new_node\n",
        "else:\n",
        "  color_dict = {'query': ['Guide gene', 'Gold', 'vee'], 'others': ['Other genes', 'Silver', 'triangle']}\n",
        "  for key in color_dict:\n",
        "    bin_name, color, shape = color_dict[key][0], color_dict[key][1], color_dict[key][2]\n",
        "    #Add node to elements\n",
        "    # { data: { id: 'a', color: 'x', shape: 'z' } },\n",
        "    new_node = \"\\t\"*5 + \"{\\n\" + \"\\t\"*6 + \"data: {\\n\" +\\\n",
        "        \"\\t\"*6 + \"id: \\'\" + bin_name + \"\\',\\n\" +\\\n",
        "        \"\\t\"*6 + \"color: \\'\" + color + \"\\',\\n\" +\\\n",
        "        \"\\t\"*6 + \"shape: \\'\" + shape + \"\\'\\n\" + \"\\t\"*6 + \"}\\n\" + \"\\t\"*5 + \"},\\n\"\n",
        "    nodes_elements += new_node\n",
        "\n",
        "legend_html_head = '''\\\n",
        "  <!doctype html>\n",
        "  <html>\n",
        "    <head>\n",
        "        <title>Network on Cytoscape</title>\n",
        "        <script src='https://cdnjs.cloudflare.com/ajax/libs/cytoscape/3.9.4/cytoscape.min.js'></script>\n",
        "    </head>\n",
        "\n",
        "    <style>\n",
        "        #cy {\n",
        "            width: 900px;\n",
        "            height: 950px;\n",
        "            }\n",
        "    </style>\n",
        "    \n",
        "    <body>\n",
        "      <div id=\"cy\"></div>\n",
        "      <script>\n",
        "        var cy = cytoscape({\n",
        "          container: document.getElementById('cy'),\n",
        "            elements: [\n",
        "\t\t\t\t          // nodes\n",
        "'''\n",
        "\n",
        "legend_html_mid ='''\n",
        "\t\t\t      ],\n",
        "            style: [\n",
        "              {\n",
        "              // Set the default node styling first\n",
        "              selector: 'node',\n",
        "              style: {\n",
        "                  shape: 'data(shape)',\n",
        "                  'height': '20px',\n",
        "                  'width': '20px',\n",
        "                  'background-color': 'data(color)',\n",
        "                  'background-opacity':1,\n",
        "                  label: 'data(id)',\n",
        "                  'border-style':'solid',\n",
        "                  'border-color':'black',\n",
        "                  'border-width':'0.1px',\n",
        "                  'font-size': 12,\n",
        "                  'text-margin-y':'-20px',\n",
        "                  'text-outline-color':'white',\n",
        "                  'text-outline-width':'0.15px',\n",
        "                  'text-outline-opacity':0.8\n",
        "                  }\n",
        "              },\n",
        "              {\n",
        "              selector: 'edge',\n",
        "              style: {\n",
        "                  'width': '0.1px',\n",
        "                  'line-style':'solid',\n",
        "                  'line-fill':'solid',\n",
        "                  'background-color': '#F5F5F5',\n",
        "                  'opacity':0.4\n",
        "                  }\n",
        "              },\n",
        "              {\n",
        "              selector:'core',\n",
        "              style: {\n",
        "                  'outside-texture-bg-color':'white',\n",
        "                  'outside-texture-bg-opacity':1\n",
        "                  }\n",
        "              }\n",
        "              ],\n",
        "            layout: {\n",
        "              name: 'grid'              \n",
        "            }            \n",
        "          });\n",
        "          cy.zoomingEnabled(false);          \n",
        "      </script>\n",
        "    </body>\n",
        "  </html>\n",
        "'''\n",
        "\n",
        "if is_a_plant == True:\n",
        "  legend_html = legend_html_head + nodes_elements[:-2] + \"\\n\" + legend_html_mid\n",
        "else:\n",
        "  legend_html = legend_html_head.replace(\"height: 950px\",\"height: 200px\").replace(\"width: 900px\", \"width: 800px\")\\\n",
        "  + nodes_elements[:-2] + \"\\n\" + legend_html_mid\n",
        "display(HTML(legend_html))"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "cd-ZU0d_ZfLB",
        "colab_type": "code",
        "cellView": "form",
        "colab": {}
      },
      "source": [
        "#@markdown 2.14 Details of Genes in Network\n",
        "#neighbourhood_file_path = neighbourhood_files[0]\n",
        "df = pd.read_table(neighbourhood_file_path, sep=\"\\t\", header=1)\n",
        "pd.set_option('display.max_colwidth', -1)\n",
        "if is_a_plant == True:\n",
        "  new_df = df[[\"Gene ID\", \"PCC Value\", \"Mapman Bin Code\", \"Mapman Bin Name\", \"Description\"]]\n",
        "else:\n",
        "  new_df = df[[\"Gene ID\", \"PCC Value\", \"Description\"]]\n",
        "new_df.style.set_properties(**{'text-align': 'left'})\n",
        "new_df.style.set_table_styles([ dict(selector='th', props=[('text-align', 'left')] ) ])"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "pOo7EKXjG3bD",
        "colab_type": "text"
      },
      "source": [
        "\\- End of Script -"
      ]
    }
  ]
}